### PR TITLE
Add org.freedesktop.LinuxAudio.BaseExtension

### PR DESCRIPTION
**Pull request by hfiguiere**
This is a base app to allow building Linux Audio app with support for plugins.

After this lands,
PR #831 will be rebased to use this as a base
PR #1459 will be build against this (renamed)
Muse3 will be modified to allow plugins.
PR #821 will be rebased too
I have a certain number of plugins to submit.

I chose `org.freedesktop.LinuxAudio` as the base for the ID.
I have tested with LV2, DSSI and LASPA, using them with Ardour (PR #831, LADSPA and LV2) and Muse3 (already in Flathub with DSSI and LADSPA)

Future development include adding VST support and supporting more applications on Flathub.

**Update**:

Following some discussions, I removed most of the modules in the Base App making it much smaller.

Here is the summary of the current state:
1. ~~The base app only include DSSI and LV2, as the headers are needed to build plugins~~
2. It declares the extension points for LADSPA, DSSI, LV2, VST (Linux native) and VST3.
3. It is based on the Freedesktop SDK instead of KDE

Caveats:
1. Carla UI that uses PyQt5 doesn't load in, say, Ardour, because it doesn't have Qt5.
2. All the plugins that use Gtk2 need to build it.

https://discourse.flathub.org/t/proposal-for-audio-plugins/399/

**Update2**:

Change the directory for each extension point as to ensure they are all separated.

**Update3**:

Just to clarify, the app don't even need to see this. It is solely to build the extension, because the extension need to thave the extension points declared at build time, hence the id change to `org.freedesktop.LinuxAudio.BaseExtension`.

**Update4**:

It includes no module.
Change the extension package id to match the path: Plugin**s** (plural)

### CODE REVIEW

**Comment by hfiguiere** [comment]
bot, build org.freedesktop.LinuxAudio.BaseApp

**Comment by hfiguiere** [comment]
Caveat: if the audio app effectively require the GNOME SDK it's a problem.

**Comment by hfiguiere** [comment]
bot, build org.freedesktop.LinuxAudio.BaseApp

**Comment by hfiguiere** [comment]
The discussion on discourse:

https://discourse.flathub.org/t/proposal-for-audio-plugins/399

**Comment by hfiguiere** [comment]
bot, build org.freedesktop.LinuxAudio.BaseApp

**Comment by hfiguiere** [comment]
Updated the description to reflect on these new changes.

**Comment by hfiguiere** [comment]
This last update assign a mount point for each type of extension.

**Comment by hfiguiere** [comment]
It has been renamed to `BaseExtension` at the suggestion of @TingPing and @barthalion

**Comment by LeandroStanger** [comment]
bot, build org.freedesktop.LinuxAudio.BaseExtension

File: org.freedesktop.LinuxAudio.BaseExtension.json [hunk 1]
```diff
@@ -0,0 +1,84 @@
+{
+  "id" : "org.freedesktop.LinuxAudio.BaseExtension",
+  "branch": "20.04",
+  "runtime" : "org.freedesktop.Platform",
+  "sdk" : "org.freedesktop.Sdk",
+  "runtime-version" : "19.08",
+  "finish-args": [
```
Comment (barthalion) [inline org.freedesktop.LinuxAudio.BaseExtension.json:7]: I don't think `finish-args` are inherited by applications using baseapps, unless you put it here for reference.

File: org.freedesktop.LinuxAudio.BaseExtension.json [hunk 2]
```diff
@@ -0,0 +1,84 @@
+{
+  "id" : "org.freedesktop.LinuxAudio.BaseExtension",
+  "branch": "20.04",
```
Comment (barthalion) [inline org.freedesktop.LinuxAudio.BaseExtension.json:3]: I'd rather have it match runtime version it targets for my own sanity.

File: org.freedesktop.LinuxAudio.BaseExtension.json (hunk 1 above)
Comment (hfiguiere) [reply org.freedesktop.LinuxAudio.BaseExtension.json:7]: Apps don't use the base app at all. This is only to build plugins.

Otherwise they are here for reference but maybe I should just put that verbatim in the README that act as a documentation for now.

Comment (barthalion) [reply org.freedesktop.LinuxAudio.BaseExtension.json:7]: Yes, it was a mental leap, I meant plugins. +1 for putting it into README.

**Comment by hfiguiere** [comment]
> As much as I dislike the idea,

What do you dislike? The idea of plugins? Or the way we (I) are doing it?

For the first, it's about making Flatpak better for the users. Audio plugins is part of the audio (music, but not only) apps. I have Ardour waiting in a pull request mostly because without audio plugin it lose a lot of versatility.

For the second, I think after many the iterations, this is the best we can do. None of the plugins that build against it do link against any of the libraries (DSSI or LV2).

I just want to make sure I address every concern.

**Comment by barthalion** [comment]
In general I'm not a fan of additional abstraction layers over shared-modules. IMO the main advantage of this base app is somewhat reduced maintenance burden although not by much really, given you will need to manually trigger builds of all plugins every time you update it. Plugins could just include the two shared modules directly, without additional base app, even if they don't link against the libraries.

**Comment by bilelmoussaoui** [comment]
> In general I'm not a fan of additional abstraction layers over shared-modules. IMO the main advantage of this base app is somewhat reduced maintenance burden although not by much really, given you will need to manually trigger builds of all plugins every time you update it. Plugins could just include the two shared modules directly, without additional base app, even if they don't link against the libraries.

I'm not a fan of the extra abstraction layer neither. I would prefer having each app to explicitly mention the various extensions that it needs/accepts instead of having that in a BaseApp.

**Comment by hfiguiere** [comment]
> Plugins could just include the two shared modules directly, without additional base app, even if they don't link against the libraries.

I doubt they'll get updated often since it's the plugin interface, even then there should be no need to recompile anything. For example some LV2 plugins don't even need it.
As for the requirement for a base app, I have already said elsewhere there seem to be no other way since the extension points need to be defined to build the Flatpak extension. The reason I included both LV2 for example is because otherwise we would have to cut and paste the module code since the shared-module has to hardcode the `/app` prefix: LV2 is built using `waf`.

**Comment by hfiguiere** [comment]
> I'm not a fan of the extra abstraction layer neither. I would prefer having each app to explicitly mention the various extensions that it needs/accepts instead of having that in a BaseApp.

As I said previously (and I update the description to clarify this), the app doesn't see this. It is solely to build the extensions. And even if then the app would still have to declare the extension points, set the environment for the path and do the install of the extension path. See #831 for an (almost) up to date example that work.

**Comment by barthalion** [comment]
> I doubt they'll get updated often since it's the plugin interface, even then there should be no need to recompile anything. For example some LV2 plugins don't even need it.

What I mean is: what's the advantage over using shared-modules? The last test build finished in less than one minute per architecture. Applications (here: extensions) don't inherit any metadata.

> because otherwise we would have to cut and paste the module code since the shared-module has to hardcode the `/app` prefix: LV2 is built using `waf`.

I don't think I understand. Why would we need to do that and why `/app` is a problem? The further we go into this discussion, the less I understand the rationale.

**Comment by hfiguiere** [comment]
> I don't think I understand. Why would we need to do that and why `/app` is a problem? The further we go into this discussion, the less I understand the rationale.

TBF I don't think I understand how I failed, but I have it working now. So I removed these two modules to. I'll need to update `shared-modules` for the rest.

Sorry about that, this should be better now.

File: org.freedesktop.LinuxAudio.BaseExtension.json (hunk 2 above)
Comment (hfiguiere) [reply org.freedesktop.LinuxAudio.BaseExtension.json:3]: done (at the next update)

**Comment by hfiguiere** [comment]
bot, build org.freedesktop.LinuxAudio.BaseExtension

**Comment by barthalion** [comment]
Now I'm completely lost. :(

Does everything this base app/ext provide is bunch of directories to mount extensions to?

**Comment by hfiguiere** [comment]
tl;dr answer: yes.

Don't worry I'm making a big write up on how do that after.

To build an extension, you need extension point. Usually an extension is for an app (Gimp, Inkscape, albeit neither handle extension flatpak, I might do that next) or a SDK. So, that app or SDK do define the extension point(s).

Audio plugins are not specific to an app as there are 5 "standard": LADSPA, DSSI, LV2, VST and VST3. So this BaseExtension "app" provide the definition of the extension points, and the audio plugins will be build against this as to remain neutral.

Application that can use these audio plugins (not all format) will need to add the extension points in their manifests, and that's it. Originally, I was thinking of providing more dependencies but it was decided it's a bad idea.

Applications on Flathub with which I have tested this:

- LMMS
- MuseSequencer
- Hydrogen
- gnusequencer
- Audacity
- Kdenlive
- Flowblade

I have all five types of plugins, built and ready to submit. Notice how two of these apps are video editors :-) And there are a couple I haven't checked yet.

Apps currently in submission state (from me) with which I have also tested:

- Ardour (the current pull request actually support Plugins now)
- Rosegarden

**Comment by hfiguiere** [comment]
I forgot, I also tested with Mixxx.

**Comment by barthalion** [comment]
I still don't understand what the BaseExtension is trying to solve here. As it does not provide dependencies, and cannot enforce any flatpak-builder settings, it's essentially useless.

Taking `org.freedesktop.LinuxAudio.Lv2Plugins.SoSynthLV2` as an example, it still needs proper `prefix` and `append-pkg-config-path` set, so it can just `mkdir -p` missing directories as well. The only upside is that `${FLATPAK_DEST}` points to correct location (I think?) but I don't consider it a good reason to add virtually empty ref which just complicates build process.

**Comment by hfiguiere** [comment]
As I said, it is still needed to define the extension points.
If I set the runtime to `org.freedeskop.Platform` then I get:

```
error: No extension point matching org.freedesktop.LinuxAudio.Lv2Plugins.SoSynthLV2 in runtime/org.freedesktop.Platform/x86_64/19.08
```

And I can't add `add-extensions` in the extension manifest to make it to work.

**Comment by hfiguiere** [comment]
Is there anything I need to do here to get this approved? Any alternative?

It is basically blocking everything else. This would be a huge step to make Linux Audio apps more viable with flatpak. And also video editors (some of them support LADSPA)

**Comment by LeandroStanger** [comment]
**Use the command:**
```
flatpak update --appstream && flatpak update -y
```

**Comment by hfiguiere** [comment]
> **Use the command:**
> 
> ```
> flatpak update --appstream && flatpak update -y
> ```

Not sure what you are responding at here and what this is supposed to fix.

**Comment by barthalion** [comment]
Sorry, I was busy. Yes, after re-reading your last two answers I think it clicked for me. Please put the rationale into README at some point. Do you have a WIP manifest enabling a plugin in some app?

**Comment by hfiguiere** [comment]
PR #831 has everything built-in. I have tested.

I have a few other apps ready to get proper support, but I was waiting for this to submit PR.

The README already mentions it, but I'll make sure it's clearer.

**Comment by hfiguiere** [comment]
Note: The ardour manifest need adjustment for the extension points, it shouldn't have "autodelete" set to true for extensions.

**Comment by barthalion** [comment]
Honestly, I'm a little bit surprised extensions point from a base app spill over like this to the runtime.

/merge

**Comment by hfiguiere** [comment]
They don't. We have to cut and paste for each app supporting them.

The apps actually don't know about this base app at all. Only the plugins.
